diff -ur nginx-1.11.7-wolfssl/src/event/ngx_event_openssl.c nginx-1.11.7-wolfssl-debug/src/event/ngx_event_openssl.c
--- nginx-1.11.7-wolfssl/src/event/ngx_event_openssl.c	2018-02-02 14:27:14.826866032 +1000
+++ nginx-1.11.7-wolfssl-debug/src/event/ngx_event_openssl.c	2018-02-02 14:28:01.823132284 +1000
@@ -134,6 +134,11 @@
 
 #endif
 
+#ifdef WOLFSSL_NGINX
+    /* Turn on internal wolfssl debugging to stdout */
+    wolfSSL_Debugging_ON();
+#endif
+
 #if OPENSSL_VERSION_NUMBER >= 0x0090800fL
 #ifndef SSL_OP_NO_COMPRESSION
     {
