diff -ur nginx-1.10.3-wolfssl/src/event/ngx_event_openssl.c nginx-1.10.3-wolfssl-debug/src/event/ngx_event_openssl.c
--- nginx-1.10.3-wolfssl/src/event/ngx_event_openssl.c	2017-04-13 10:38:27.618124865 +1000
+++ nginx-1.10.3-wolfssl-debug/src/event/ngx_event_openssl.c	2017-04-13 11:28:45.139830218 +1000
@@ -128,6 +128,11 @@
 
 #endif
 
+#ifdef WOLFSSL_NGINX
+    /* turn on internal wolfssl debugging to stdout */
+    wolfSSL_Debugging_ON();
+#endif
+
 #if OPENSSL_VERSION_NUMBER >= 0x0090800fL
 #ifndef SSL_OP_NO_COMPRESSION
     {
