diff --git a/src/event/ngx_event_openssl.c b/src/event/ngx_event_openssl.c
index 4d69557b6..1c1ea2c19 100644
--- a/src/event/ngx_event_openssl.c
+++ b/src/event/ngx_event_openssl.c
@@ -134,6 +134,13 @@ int  ngx_ssl_certificate_name_index;
 
 u_char  ngx_ssl_session_buffer[NGX_SSL_MAX_SESSION_SIZE];
 
+#ifdef WOLFSSL_NGINX
+void wolfSSL_Log(const int logLevel, const char *const logMessage)
+{
+    (void)logLevel;
+    ngx_log_stderr(0, "wolfSSL: %s", logMessage);
+}
+#endif
 
 ngx_int_t
 ngx_ssl_init(ngx_log_t *log)
@@ -208,6 +215,12 @@ ngx_ssl_init(ngx_log_t *log)
    #endif
 #endif
 
+#ifdef WOLFSSL_NGINX
+    /* Turn on internal wolfssl debugging to stderr */
+    wolfSSL_SetLoggingCb(wolfSSL_Log);
+    wolfSSL_Debugging_ON();
+#endif
+
 #ifndef SSL_OP_NO_COMPRESSION
     {
     /*
