diff -ur nginx-1.11.7-wolfssl/src/event/ngx_event_openssl.c nginx-1.11.7-wolfssl-debug/src/event/ngx_event_openssl.c
--- nginx-1.11.7-wolfssl/src/event/ngx_event_openssl.c	2017-01-18 16:30:54.859646118 +1000
+++ nginx-1.11.7-wolfssl-debug/src/event/ngx_event_openssl.c	2017-04-13 11:27:49.123757568 +1000
@@ -134,6 +134,12 @@
 
 #endif
 
+#ifdef WOLFSSL_NGINX
+    /* turn on internal wolfssl debugging to stdout */
+    wolfSSL_Debugging_ON();
+#endif
+
+
 #if OPENSSL_VERSION_NUMBER >= 0x0090800fL
 #ifndef SSL_OP_NO_COMPRESSION
     {
