diff -ur nginx-1.17.5-wolfssl/src/event/ngx_event_openssl.c nginx-1.17.5-wolfssl-debug/src/event/ngx_event_openssl.c
--- nginx-1.17.5-wolfssl/src/event/ngx_event_openssl.c	2019-11-04 21:29:39.856200843 +0100
+++ nginx-1.17.5-wolfssl-debug/src/event/ngx_event_openssl.c	2019-11-04 21:30:29.362010122 +0100
@@ -165,6 +165,12 @@
 
 #endif
 
+#ifdef WOLFSSL_NGINX
+    /* Turn on internal wolfssl debugging to stdout */
+    wolfSSL_Debugging_ON();
+#endif
+
+
 #ifndef SSL_OP_NO_COMPRESSION
     {
     /*
