diff -ur nginx-1.11.10-wolfssl/src/event/ngx_event_openssl.c nginx-1.11.10-wolfssl-debug/src/event/ngx_event_openssl.c
--- nginx-1.11.10-wolfssl/src/event/ngx_event_openssl.c	2017-04-13 14:53:51.151297965 +1000
+++ nginx-1.11.10-wolfssl-debug/src/event/ngx_event_openssl.c	2017-04-13 15:43:18.269591752 +1000
@@ -144,6 +144,11 @@
 
 #endif
 
+#ifdef WOLFSSL_NGINX
+    /* Turn on internal wolfssl debugging to stdout */
+    wolfSSL_Debugging_ON();
+#endif
+
 #if OPENSSL_VERSION_NUMBER >= 0x0090800fL
 #ifndef SSL_OP_NO_COMPRESSION
     {
