diff -ur nginx-1.12.0-wolfssl/src/event/ngx_event_openssl.c nginx-1.12.0-wolfssl-debug/src/event/ngx_event_openssl.c
--- nginx-1.12.0-wolfssl/src/event/ngx_event_openssl.c	2017-04-13 09:54:22.594624149 +1000
+++ nginx-1.12.0-wolfssl-debug/src/event/ngx_event_openssl.c	2017-04-13 11:29:35.243899163 +1000
@@ -144,6 +144,11 @@
 
 #endif
 
+#ifdef WOLFSSL_NGINX
+    /* turn on internal wolfssl debugging to stdout */
+    wolfSSL_Debugging_ON();
+#endif
+
 #if OPENSSL_VERSION_NUMBER >= 0x0090800fL
 #ifndef SSL_OP_NO_COMPRESSION
     {
